Expand Stylguide (#463)
clerk/clerk-docs: STYLEGUIDE.md

#### Don't 

> The [`currentUser()`](https://clerk.com/docs/references/nextjs/current-user) helper will return the [`User`](https://clerk.com/docs/references/javascript/user/user) object of the currently active user. The following example userse the [`currentUser()`](https://clerk.com/docs/references/nextjs/current-user) helper to access the [`User`](https://clerk.com/docs/references/javascript/user/user) object for the authenticated user.

## 2. Write for clarity

### 2.1 “We” are the authors; “you” is the reader.

Clerk is the entity writing these docs and providing these services. The learner is the person consuming these things and building their own projects. To distinguish between the two and write with clarity, we write from Clerk's perspective, using "we/us/our/ours," and refer to the learner with "you/your/yours."

#### Do

> Our `<ClerkProvider />` provides active session and user context to Clerk's hooks and other components. Import it into your app by adding `import { ClerkProvider } from '@clerk/nextjs'` at the top of your file.

#### Don't 

> > `<ClerkProvider />` provides active session and user context to our hooks and other components. Let's import it into our app by adding `import { ClerkProvider } from '@clerk/nextjs'` at the top of the file.

Reviewer: Would like @BRKalow input on this - we had discussed the use of "we/our/ours" should not be used in the docs, and instead use objective language when referring to Clerk.
Instead of "Our `<ClerkProvider />`", it would be written "Clerk's `<ClerkProvider/ >`".

Author: I'm so glad you've been thinking about this! Want to get alignment before writing this in stone.

Reviewer 2: I feel like we/us/our/ours works for blog posts, Twitter and things like that, while the Docs should be more 'Clerk.' I just took a look at the Next.js and Prisma sites and they follow that convention --- we/us/our/ours in the Blog Pots and Next.js/Prisma in the docs. That said, I think either works in the docs and its more a matter of revising to make that consistent.

Reviewer 3: That's right, up to this point we've tried to use `Clerk` instead of the mentioned pronouns. I personally find it a more effective writing style for reference-like material. Like @royanger said, we/us/our/ours feels more appropriate in blog posts and other mediums where there is generally a clear individual author listed.

Ultimately I think consistency should win here, so whatever recommendation we go with is okay with me as long as we're consistent. 👍

Author: This styleguide seeks to pave cowpaths rather than enforce new voice and tone. If this pattern were harmful, we might change it, but it makes sense, and it sounds like our docs already use it. Let's make this the official recommendation and canonize it in the styleguide

### 2.2 "Users" are people logging in/out with Clerk. "Developers" are the audience we’re writing for.

> People building with clerk are building for other people to make it clear whether we are talking about the builder or their users, we call builders "developers" and the peoeple they are building for "users."

Reviewer: ```suggestion
People building with Clerk are building for other people. To make it clear whether we are talking about the builder or their users, we call builders "developers" and the people they are building for "users."
```

#### Do

> Clerk is now installed and mounted in your application. You can decide which pages are public or require authentication to access.

#### Don't

> Now that Clerk is installed and mounted in your application, you can decide which pages are public and which should require authentication to access.

### 2.4 Avoid gerunds (-ing words)

English gerunds ("-ing" words like "running") turn verbs into nouns ("run" becomes "running"). This makes the sentence sound passive ("They run" becomes "they are running") and makes it harder to translate. Use an active voice as much as possible and avoid these words.

#### Do

> How to use a routing library with Clerk

#### Don't

> Using a routing library with Clerk

### 2.5 Use task and outcome-based headings and titles that use verbs.

Tell the learner what they will learn from diving into the content. 

#### Do

> How to read session and user data

> How user sessions and user data work

Clearly explains what information the learner will find, helping them decide which to choose.

#### Don't

> User session and user data

> Obscure. Doesn't convey whether this is the conceptual or task-based content the learner is looking for. They may pass over it or look briefly, then bounce out. This increases friction.

Reviewer: ```suggestion
Obscure. Doesn't convey whether this is the conceptual or task-based content the learner is looking for. They may pass over it or look briefly and then leave. This increases friction.
```

### 2.6 Lead with location; end with action.

When learners are performing an order of operations, it helps for them to start with _where_ they need to be to perform the action.

#### Do

> In your Next.js project's root folder, open your `.env.local file`.

> The learner navigates to the project folder first, opens the file second.

Reviewer: ```suggestion
The learner navigates to the project folder first, and opens the file second.
```

#### Don't

> Open your `.env.local file` in your Next.js project's folder.

The learner must remember the file they need to open before finding out which folder to navigate to. This uses more mental energy.

### 2.7 Avoid colloquialisms

> English speakers my find themselves using phrases like "

Reviewer: This is missing a phrase. The quotation is just hanging here.

```suggestion
English speakers may find themselves using phrases like "
```

#### Do

> Something that is difficult for someone new to programming may not be for a senior engineer.

#### Don't

> Something that is difficult for someone new to programming may be child's play for a senior engineer.

Some learners may not have heard this expression before. It may be difficult for machine translation as well.

## 3. Write to include people

### 3.1 Do not assume proficiency

Avoid using language that assumes someone's level of proficiency. Something that is difficult for someone new to programming may not be difficult for a senior engineer. This language can inadvertently alienate or insult a learner. Avoid words like "just," "easy," "simple," "senior," "hard."

#### Do

> > You can authenticate your app with Clerk in three steps. Install Clerk with `npm install @clerk/nextjs`, add your environment keys, and then wrap your app in `<ClerkProvider />`, and add [control components](https://clerk.com/docs/components/overview). Visit our [Quickstarts](https://clerk.com/docs/quickstarts/overview) for a step-by-step guide written for your framework.

Reviewer: ```suggestion
> You can authenticate your app with Clerk in three steps. Install Clerk with `npm install @clerk/nextjs`, add your environment keys, and then wrap your app in `<ClerkProvider />`. Visit our [Quickstarts](https://clerk.com/docs/quickstarts/overview) for a step-by-step guide written for your framework.
```

Author: We're using "control components" somewhere... Are we not using that terminology anymore?

#### Do

> When the user logs out using the button, they will be taken to the login page.

#### Don't

> When he logs out using the button, he will be taken to the login page.

### 3.4 Avoid "click"

"Click" is an outdated term that assumes the learner is using a mouse. But learners may be navigating by touchscreen, keyboard, or assistive technology. Often there are better words than "click"—like "select" and "open".

#### Do

> Open the **Settings tab.**

> Select the **Google** social connection.

#### Don't

> Click the **Settings tab.** 

> Click the **Google** social connection.

## 4. Write for code

### 4.1 Use monospace fonts for code, commands, file names, copy-ables, and URLs; use bold for menus, and other UI copy.

#### Do

> In your [Clerk Dashboard](https://dashboard.clerk.com/), go to **User & Authentication > Social Connections** and open the **Settings tab.**

> In your browser, open [`http://localhost:3000/`](http://localhost:3000/).

#### Don't

> In your [Clerk Dashboard](https://dashboard.clerk.com/), go to `User & Authentication > Social Connections` and open the "Settings tab."

> In your browser, open [`http://localhost:3000/`](http://localhost:3000/).

> ### 4.2 Use carets to help users navigate menus; 

Reviewer: ```suggestion
### 4.2 Use carets to help users navigate menus
```

Use carets to nest operations of the same type like menu navigation.

#### Do

> In your [Clerk Dashboard](https://dashboard.clerk.com/), go to **User & Authentication > Social Connections** and open the **Settings tab.**

#### Don't

> In your [Clerk Dashboard](https://dashboard.clerk.com/), go to **User & Authentication**, click on **Social Connections**, and open the **Settings tab.**

### 4.3 Write compliant JSON: no single quotes, no comments

Conform to the [JSON standard](https://www.json.org/json-en.html):

* Key names are enclosed in double quotes (`"`)
* Values are between double quotes (`"`), not single quotes (`'`)
* No comments, neither `//` nor `/* */`

#### Do

> This is the user's primary email address in the User object

```json
{
  "email": "{{user.primary_email_address}}"
}
```

#### Don't

```json
{
	// this is the user's primary email address in the User object
  email: '{{user.primary_email_address}}'
}
```
 
### 4.4 Use 2-space indents for code blocks

> We use 2-space indents to conserve horizontal space in code blocks.

Reviewer 4: Can we enforce this with automation or tooling?

Reviewer 2: I would love to have or help setup a prettier that can be used to format code before adding it to dpcs.

Author: That's the plan, Roy! As soon as this PR is merged/approved by all, we can start linting for as many of these as possible.

Reviewer 2: Nice! I would also propose making them available in the repo if possible, so contributors can use (if that is not part of the plans already)

#### Do

```js
export const config = {
  matcher: ['/((?!.+\\.[\\w]+$|_next).*)', '/', '/(api|trpc)(.*)'],
};
```

#### Don't

```js
export const config = {
    matcher: ['/((?!.+\\.[\\w]+$|_next).*)', '/', '/(api|trpc)(.*)'],
};
```

### 4.5 Authentication states should be hyphenated, bold, and use monospace fonts.

Clerk authentication states **`signed-in`**, **`signed-out`**, and **`unknown`** should be hyphenated, bold, and use monospace fonts to distinguish them from general states of being.

#### Do

> Once you have a **`signed-in`** user, you need to give them a way to sign out.

#### Don't

> > Once you have a signed-in user, you need to give them a way to sign out.

Reviewer: I don't remember this being a discussion, and I truly believe this is going to create a lot of visual noise in the docs. Code snippets should be saved for code-related copy. Saying a user is signed-in is not necessarily code-related. I think if we want to bold the term, maybe that could be a feasible solution. I do not support using bold AND code snippets, nor code snippets at all, to distinguish signed-in, signed-up, signed-out states.

Reviewer 2: I think I side with Alexis on this. I feel like bold is a good choice. Using backticks could feel out of place and confusing to developers reading the docs.

Author: Bold it is! Thanks for reviewing and adding feedback!
